index.js

```diff
@@ -1,122 +1,93 @@
-            e[f++] = l >> 18 | 240,
-            e[f++] = l >> 12 & 63 | 128) : e[f++] = l >> 12 | 224,
-            e[f++] = l >> 6 & 63 | 128),
-            e[f++] = l & 63 | 128)
-    }
-    a = b;
-    for (f = 0; f < e.length; f++)
-        a += e[f],
-            a = xr(a, "+-a^+6");
-    a = xr(a, "+-3^+b+-f");
-    a ^= Number(d[1]) || 0;
-    0 > a && (a = (a & 2147483647) + 2147483648);
-    a %= 1E6;
-    return c + (a.toString() + "." + (a ^ b))
-}
-
-var yr = null;
-var wr = function(a) {
-    return function() {
-        return a
-    }
-}
-    , xr = function(a, b) {
-    for (var c = 0; c < b.length - 2; c += 3) {
-        var d = b.charAt(c + 2)
-            , d = "a" <= d ? d.charCodeAt(0) - 87 : Number(d)
-            , d = "+" == b.charAt(c + 1) ? a >>> d : a << d;
-        a = "+" == b.charAt(c) ? a + d & 4294967295 : a ^ d
-    }
-    return a
-};
+var config = new Configstore('google-translate-api');
+var batchNumber = 0;
 
-// END
-/* eslint-enable */
 
-var config = new Configstore('google-translate-api');
+/**
+ * @param {string} rpcId                    ID of service to call
+ * @param {Array} payload                   Arguments for the service call
+ * @param {string} [opts.tld="com"]         translate.google.[tld]
+ * @param {number} [opts.tokensTTL=3600]    How long to cache tokens
+ */
+function batchExecute(rpcId, payload, opts) {
+    if (!opts) opts = {};
+    if (!opts.tld) opts.tld = "com";
+    if (!opts.tokensTTL) opts.tokensTTL = 3600;
 
-var window = {
-    TKK: config.get('TKK') || '0'
-};
+    var url = "https://translate.google." + opts.tld;
 
-function updateTKK() {
-    return new Promise(function (resolve, reject) {
-        var now = Math.floor(Date.now() / 3600000);
+    return Promise.resolve(config.get("wiz"))
+        .then(function(wiz) {
+            if (wiz && (wiz.timestamp + opts.tokensTTL * 1000) > Date.now()) return wiz;
+            return fetchWizGlobalData(url)
+                .then(function(wiz) {
+                    wiz.timestamp = Date.now();
+                    config.set("wiz", wiz);
+                    return wiz;
+                })
+        })
+        .then(function(wiz) {
+            return getBatchExecuteParams(wiz, rpcId, payload);
+        })
+        .then(function(params) {
+            if (!params.body.at) delete params.body.at;
+            return got.post(url + "/_/TranslateWebserverUi/data/batchexecute", {
+                    searchParams: params.query,
+                    form: params.body,
+                    responseType: "text"
+                })
+                .then(function(res) {
+                    var match = res.body.match(/\d+/);
+                    return res.body.substr(match.index + match[0].length, Number(match[0]));
+                })
+                .then(JSON.parse)
+                .then(function(envelopes) {
+                    var payload = envelopes[0][2];
+                    return JSON.parse(payload);
```

Comment:
He always returns null for some reason Inconsistent with tk parameter

Reply:
Yes, if you pass incorrect values for the rpc-id or the payload, it will return null.  Really unhelpful indeed.

Try the following, should return valid payload: `batchExecute("jQ1olc", ["Hello, world!", "en", null])`